File: 6322-497-802-find-eventual-safe-states/6322-497-802-find-eventual-safe-states.py
class Solution:
    def dfs(self, u, graph, state):
        if state[u] != 0:
            return state[u] == 1
        state[u] = 2
        for child in graph[u]:
            if not self.dfs(child, graph, state):
                return False
        state[u] = 1
        return True

    def eventualSafeNodes(self, graph):
        n = len(graph)
        state = [0] * n
        ans = []
        for i in range(n):
            if self.dfs(i, graph, state):
                ans.append(i)
        return ans

File: 6322-497-802-find-eventual-safe-states/6322-497-802-find-eventual-safe-states.js
var eventualSafeNodes = function(graph) {
    const n = graph.length;
    const state = Array(n).fill(0); 
    const ans = [];
    
    const dfs = (u) => {
        if (state[u] !== 0) return state[u] === 1;
        state[u] = 2;
        for (const child of graph[u]) {
            if (!dfs(child)) return false;
        }
        state[u] = 1;
        return true;
    };

    for (let i = 0; i < n; i++) {
        if (dfs(i)) ans.push(i);
    }
    return ans;
};
